### ISSUE TITLE
Limit decorator for non standard routing (Flask-Restful)

### ISSUE THREAD

**Issue by marijns** [alisaifee/flask-limiter#11]
Current situation:
When the decorator is loaded, a route is added to route_limits or dynamic_route_limits. The name of the route is defined like this:

``` python
name = "%s.%s" % (fn.__module__, fn.__name__)
```

Then for every 'before request', __check_request_limit is called. This function tries to find the correct name for the end point and see if there is a limit defined for it in either dictionaries:

``` python
endpoint = request.endpoint or ""
        view_func = current_app.view_functions.get(endpoint, None)
        name = ("%s.%s" % (
                view_func.__module__, view_func.__name__
            ) if view_func else ""
```

This works well for standard routing. But now I want to use Flask-Limiter for Flask-Restful routes. And this creates a problem because the endpoint name when _checking_ the limit will be different than the name that is determined when the decorator is _defined_.

For example, module myapp.api:

``` python
class HelloWorld(restful.Resource):
    @limiter.limit("1/minute")
    def get(self):
        return {'hello': 'world'}

api.add_resource(HelloWorld, '/')
```

would create something like myapp.api.get as name for the limiter route, but is actually defined as myapp.api.helloworld by Flask-Restful. Of course this is not specific to Flask-Restful, and can occur any time the limiter is used on a function that is not also the actual endpoint, I think.

Any thoughts on how to fix this in a nice way? Maybe the checking of the rate limiting could be better done directly in the decorator instead of 1 central before_request?

**Comment by ZhangBohan** [comment]
When I use `flask-limiter` in `Flask-Restful`, `decorators` is worked

```
decorators = [limiter.limit("2/minute")]
```
> restricted to only certain http methods by passing them as a list to the methods keyword argument.

as the doc says, I pass the method name as a list, but limiter no longer work at all, can you help me to this question?

```
decorators = [limiter.limit("2/minute", methods=['put'])]
```

**Comment by alisaifee** [comment]
@ZhangBohan : theres a bug in the code as it stands where the method name is case sensitive. If you use `methods=["PUT"]` instead it will work. I'll make a dot release in a bit to fix that issue though.

**Comment by ZhangBohan** [comment]
@alisaifee Many thanks!! It's working! :+1:

**Comment by fadziljusri** [comment]
Hi, let's say I need different limit for different methods. How can I achieve that? I try codes below but only 'GET' can have limit, but not 'POST'

```python
class Test(Resource):
 decorators = [
  limiter.limit("2/minute", methods=['GET']),  # custom limiter for GET
  limiter.limit("1/minute", methods=['POST']),  # custom limiter for POST
 ]

 def get(self):
  pass

 def post(self):
  pass
```

**Comment by kenedinovriansyah** [comment]
where did you get the decorators

**Comment by cmin764** [comment]
> Hi, let's say I need different limit for different methods. How can I achieve that? I try codes below but only 'GET' can have limit, but not 'POST'
> 
> ```python
> class Test(Resource):
>  decorators = [
>   limiter.limit("2/minute", methods=['GET']),  # custom limiter for GET
>   limiter.limit("1/minute", methods=['POST']),  # custom limiter for POST
>  ]
> 
>  def get(self):
>   pass
> 
>  def post(self):
>   pass
> ```

Yes, I can confirm this. The only workaround I found was to split the get and post into two separate resource classes and register both of them, this way you can have different rate limitings for the different methods, but this is very ugly and un-defines the purpose of having resources with (GET/POST/etc. methods).

The problem in the limiter logic is that it simply exempts the method if is not found among the enabled ones instead of recursively go through the decorators to see if there's at least one limit that applies for the given route. Like so:

@get_limit
@post_limit
def post
--> the get_limit is enabled for GETs only, therefore is exempted, while the post_limit applies, but is skipped entirely due to the previous ignored case on top of that.

# Potential solution
Simply have a class-based decorator that can check the chain of previously applied limits so it can jump to the non-exempted ones while gathering the enabled limits.

**Comment by kenedinovriansyah** [comment]
from limiter import get_limiter, limit

REFRESH_RATE = 2BURST_RATE = 3

limiter = get_limiter(rate=REFRESH_RATE, capacity=BURST_RATE)

@limit(limiter)def get_page(url: str) -> Response:
    return get(url)

@limit(limiter, consume=2)async def do_stuff():
    await sleep(0.1)

def do_stuff():
    # do stuff

    with limit(limiter, consume=1.5):
        # do expensive stuff
        pass

async def do_stuff():
    # do stuff

    async with limit(limiter):
        # do expensive stuff
        pass

@limit(limiter, bucket=b'other stuff')def do_other_stuff():
    pass

For Flask

from flask_limiter import Limiterfrom flask_limiter.util import
get_remote_address
app = Flask(__name__)limiter = Limiter(
    app,
    key_func=get_remote_address,
    default_limits=["2 per minute", "1 per second"],)
@app.route("/slow")@limiter.limit("1 per day")def slow():
    return "24"
@app.route("/fast")def fast():
    return "42"
@app.route("/ping")@limiter.exemptdef ping():
    return 'PONG'


Pada tanggal Rab, 29 Jan 2020 pukul 15.09 Cosmin Poieana <
notifications@github.com> menulis:

**Comment by alisaifee** [comment]
This is a bug in the code where a `return` statement is used instead of `continue` when one of the limits that needs to be tested is exempt for whatever reason.

**Comment by alisaifee** [comment]
@cmin764 & @fadziljusri please let me know if the above fix will solve your issues and I'll make a release.

**Comment by cmin764** [comment]
Thank you @alisaifee for your prompt action and while this patch really adds value to the library and clearly fixes some cases, it still doesn't solve our issue because that `continue` instead of `return` you introduced still skips our default limit applied over POST when we define `decorators = [...]` for GETs only.

I've done an experiment by putting `0/0` (which raises zero division exception) right before your continue like this:
```python
            if lim.methods is not None and request.method.lower(
            ) not in lim.methods:
                0/0 ####
                continue
```
and only like this, the overriding GET-only limit is bypassed, therefore finally leaving the POST limit unaffected.

I have the following API code:

```python
def _token_key_func(token, *, share):
    auth_token = base.get_auth_token()
    if auth_token == token:
        token_hash = hashlib.md5(auth_token.encode(settings.ENCODING)).hexdigest()
        ip_addr = f":{get_remote_address()}" if share else ""
        return f"{token_hash}{ip_addr}"

    return None


def _get_counter_article_limits(**kwargs):
    limits = []
    limiter_conf = app.config["CONFIG"].rate_limiter
    for email in limiter_conf.emails:
        limit_str = f"{limiter_conf.default};{email.limit}"
        token = auth.compute_token(email.email)
        key_func = functools.partial(
            _token_key_func, token, share=email.share or False
        )
        limit = limiter.limit(limit_str, key_func=key_func, **kwargs)
        limits.append(limit)
    return limits


def _exempt_when_abort():
    try:
        request.articles_pair = GetCounterArticleResource.get_related_articles()
    except HTTPException as exc:
        logging.debug("Couldn't get related articles due to: %s", exc)
        request.exception = exc
        return True
    else:
        request.exception = None

    return False


class BaseArticleResource(base.DatastoreMixin, base.BaseResource):

    """Base class for all article related resources."""

    URL_PREFIX = "/article"

    @staticmethod
    def _get_schemas():
        # Required due to circular import.
        from truestory.schemas.article import (
            article_schema, articles_schema
        )
        schemas = {
            "article": article_schema,
            "articles": articles_schema,
        }
        return schemas


class BaseCounterArticleResource(BaseArticleResource):

    ENDPOINT = "counter"

    @classmethod
    def get_route(cls):
        return super().get_route().replace(
            cls.ENDPOINT, BaseCounterArticleResource.ENDPOINT
        )


class GetCounterArticleResource(BaseCounterArticleResource):

    """Handles GETs of opposite articles."""

    ENDPOINT = f"get_{BaseCounterArticleResource.ENDPOINT}"

    if not app.debug:
        _LIM_KWARGS = {"per_method": True, "methods": ["GET"]}
        _DEFAULT_LIMIT = app.config["RATELIMIT_DEFAULT"]
        decorators = [
            limiter.limit(_DEFAULT_LIMIT, key_func=get_remote_address, **_LIM_KWARGS)
        ] + _get_counter_article_limits(exempt_when=_exempt_when_abort, **_LIM_KWARGS)

    @staticmethod
    def get_related_articles():
        data = request.get_json(silent=True) or request.args
        link = data.get("link", "").strip()
        if not link:
            abort(400, message="Article 'link' not supplied.")

        link = strip_article_link(link)
        article_query = ArticleModel.query(("link", "=", link))
        main_articles = ArticleModel.all(query=article_query, keys_only=True, limit=1)
        if not main_articles:
            abort(404, message="Article not found in the database.")

        main_article = main_articles[0]
        related_articles = ArticleModel.get_related_articles(main_article.key)
        if not related_articles:
            abort(404, message="No related articles found.")

        return main_article, related_articles

    def get(self):
        """Returns a list of opposite articles for the provided one."""
        if app.debug:
            main_article, related_articles = self.get_related_articles()
        else:
            if request.exception:
                raise request.exception
            main_article, related_articles = request.articles_pair

        articles = []
        unique_sources = set()
        for article in sorted(
                related_articles, key=operator.itemgetter("score"), reverse=True
        ):
            article = article["article"]
            if article.source_name in unique_sources:
                continue
            else:
                unique_sources.add(article.source_name)
            articles.append(article)
            if len(articles) >= settings.API_MAX_RELATED_ARTICLES:
                break

        main_article_url = self._make_response(
            "articles", [main_article]
        ).json["articles"][0]
        return self._make_response("articles", articles, main=main_article_url)


class PostCounterArticleResource(BaseCounterArticleResource):

    """Handles POSTs in opposite articles."""

    ENDPOINT = f"post_{BaseCounterArticleResource.ENDPOINT}"

    def post(self):
        data = request.get_json()
        link = data.get("link", "").strip() if data else None
        if not link:
            abort(400, message="Article 'link' not supplied.")

        link = strip_article_link(link)
        try:
            site, site_info = ArticleModel.get_site_info(link)
        except Exception as exc:
            abort(403, message=base.exc_to_str(exc))

        article = _extract_article(link, site, site_info)
        article_key = article.put()
        article_usafe = key_to_urlsafe(article_key)
        try:
            _pair_article(article_usafe)
        except Exception as exc:
            logging.exception(
                "Couldn't pair article with urlsafe '%s' due to: %s",
                article_usafe,
                exc
            )
        return self._make_response("article", article)
```

and this works wonderfully with the current official release of the library (returns or continues, doesn't matter), but to solve this matter, I just want to have both the POST and GET under the same resource and while having the default global limit available for POST, to be able to define other (overriding) limits for GET only. Unfortunately, even with your patch, the limit is exempted for POST entirely when we define GET-only limits inside a resource.

Given your patch, another workaround might be the necessity of adding another explicit redundant limit for POSTs only too, but the idea here is to just skip workarounds and have the possibility of keeping all the limits as they are for the unaffected methods (like POSTs, while explicitly defining a new limit for GETs only, like you see in our example).

Thanks a lot for your understanding and work done so far + maybe a future patch taking care of the unaffected methods in order to keep their limits.
My hunch is that the limit applied over GETs only, somehow applies to POSTs too, then the logic checks if that limit is allowed for POSTs (and it isn't given `methods` param), therefore is skipped, but in the process we somehow lose the previous default limit applied already over POST (maybe due to decorators over decorators or some strange order in applying them one over another).

Please keep in mind the fact that in my \_\_init\_\_.py file I have already defined defaults like this: `limiter = Limiter(app, key_func=get_remote_address)` while the configuration is:
```python
    RATELIMIT_DEFAULT = CONFIG.rate_limiter.default
    RATELIMIT_STORAGE_URL = misc.get_redis_url()
    RATELIMIT_HEADERS_ENABLED = True
    RATELIMIT_IN_MEMORY_FALLBACK = RATELIMIT_DEFAULT
    RATELIMIT_KEY_PREFIX = "truestory"
    RATELIMIT_SWALLOW_ERRORS = True
```
and clearly I want these limits preserved over the entire app/api and have them overridden if-and-only-if I explicitly provide such override through specific method I select given the `decorators = [...]` class attribute under the affected resource, WITHOUT requiring me to split the `get()` (affected) and `post()` (unaffected) methods into different classes like I've done in the example above.

I think this explains the whole problem.

**Comment by alisaifee** [comment]
@cmin764 thanks for the detailed feedback. Based on what @fadziljusri posted I feel the patch will probably address his issue. What you're running into is more complex. If you're open to it could you please drop me an email so we can take this offline for a short while as I'm not confident that I completely understand your problem and I'd like to move this to a new issue.

**Comment by fadziljusri** [comment]
> @cmin764 & @fadziljusri please let me know if the above fix [[934dcfe](https://github.com/alisaifee/flask-limiter/commit/934dcfef404a78b000690a33cbb81c7d72d60357)] will solve your issues and I'll make a release.

Hi @alisaifee , yes this fix my issues with flask-restful package 👍

**Comment by cmin764** [comment]
> @cmin764 thanks for the detailed feedback. Based on what @fadziljusri posted I feel the patch will probably address his issue. What you're running into is more complex. If you're open to it could you please drop me an email so we can take this offline for a short while as I'm not confident that I completely understand your problem and I'd like to move this to a new issue.

Hello and thank you for your interest. My e-mail address is: cmin764@gmail.com and Skype: cmin764

I've simplified the problem into the following script (**lim.py**) which can be run like so: `FLASK_ENV=development FLASK_APP=lim flask run -p 8080` and has the content:
```python
from flask import Blueprint, Flask
from flask_limiter import Limiter
from flask_limiter.util import get_ipaddr as get_remote_address
from flask_restful import Api, Resource


class Config:
    RATELIMIT_DEFAULT = "1/3second"
    RATELIMIT_STORAGE_URL = "redis://"
    RATELIMIT_HEADERS_ENABLED = True
    RATELIMIT_IN_MEMORY_FALLBACK = "1/2second"
    RATELIMIT_KEY_PREFIX = "test-limiter"
    RATELIMIT_SWALLOW_ERRORS = True


app = Flask(__name__)
app.config.from_object(Config)

limiter = Limiter(app, key_func=get_remote_address)

api_bp = Blueprint("api", __name__)
api = Api(api_bp)
app.register_blueprint(api_bp, url_prefix="/api")


class MyResource(Resource):

    decorators = [
       limiter.limit("3/minute", key_func=get_remote_address, methods=["GET"]),
       #limiter.limit("1/3second", key_func=get_remote_address, methods=["POST"])    # this works with your patch if I uncomment
    ]

    def get(self):
        print("GET")
        return {"m": "get"}

    def post(self):
        print("POST")
        return {"m": "post"}


api.add_resource(MyResource, "/my", endpoint="my")
```

And the requests can be done like so (with _httpie_):
- GET: `http :8080/api/my`
- POST: `http POST :8080/api/my`

And as you can see, the `"3/minute"` GET-only limit (through `decorators = [...]`) simply exempts my default POST limit, why? If I comment this limit, then my POST (and the rest of the methods work fine, as expected) and if I try to explicitly set another default limit for POST method (`"1/3second"` like seen in the 2nd commented limit) by just have uncommented both limits, the POST limit is still exempted with the current library, but given your patch, this problem is solved if-and-only-if we make that workaround for re-specifying then again an explicit limit for the rest of the methods (like POST).

Wouldn't be nice to not be obliged to explicitly specify default limits for unaffected methods? (like POSTs in my scenario)
When I say `methods=["GET"]` doesn't this mean that I want only GETs given that resource being affected?
Secondly: I had previously a `1/3second` limit for any type of request (applied over the GET too). When I specify a new limit (like in `decorators = [...]` under the resource), aren't we expecting to have that decorator put on top of the default one (having both `1/3second` default limit + `3/minute` for GETs only)? Or at least have an `override=True` flag under `.limit()` so if I switch this to `False` I'll have this limit compounded to the other already available/set ones. But this is a finer problem; would be nice to fix at least the one above: when specifying a GET-only limit to not lose the POST one (and be forced to make the workaround of specifying one for the POST too just because we wanted a custom one for GET under the same resource).

I think this example is much more clearer now, please let me know if you're still perplexed.
